README.md
## fecon235 :: Notebooks for financial economics

[![Join the chat at https://gitter.im/rsvp/fecon235](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/rsvp/fecon235?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
/ [CHANGELOG] / [BSD License and TOS][235li]

**Spin-off Notice:** ***the modules and functions used by our research notebooks
have been refactored into another repository, [fecon236].***
The home for *our Jupyter notebooks will remain here at [fecon235]*
in the `nb` directory.


### What is this repository for?

**fecon235** provides an interface for ***financial economics*** to the Python
ecosystem, especially packages for mathematics, statistics, science,
engineering, and data analysis.
Complex packages such as *numpy, pandas, statsmodels, scipy, and matplotlib*
are seamlessly integrated at a high-level with APIs of various data hosts for:

- Essential commands which correctly handle annoying low-level pitfalls.

- Retrieval of economic and financial data, both historical and the most current. 

- Data munging, for example, resampling and alignment of time-series data
  from hosts using mutually incompatible formats.

- Analysis using techniques from econometrics, time-series analysis,
  and statistical machine learning.

- Abstraction and software optimization of mathematical operators,
  for example, linear algebra used in portfolio analysis.

- Visualization of data using graphical packages. 

- *Reproducible research which is collaborative and openly accessible
  at zero cost.*

To practically test theoretical ideas interactively,
our code can employed with any Python IDE interactive development
environment, IPython console, or with a Jupyter notebook.
The code has been tested against both python27 and python3 since 2014,
and works across major platforms: Linux, Mac, and Windows.

**Database:** the primary source is [FRED], the U.S. Federal Reserve
Economic Data bank which is accessed directly online using our interface. 
Other specialized data can be directly retrieved using our [Quandl] API 
module, for example, futures prices. Data for stocks, mutual funds, and 
ETFs is retrieved from the best available sources using `pandas_datareader`.
Data is designed to be accessible *free* of charge,
and *interoperable* in different time frequencies.


### How does one get started?

* Good introductory lectures for economists to Python and its ecosystem:
  Quantitative Economics by Thomas [Sargent].

* For the fecon235 and fecon236 installation FAQ, please
  see https://git.io/econ which also covers external dependencies.

* For the older pre-2016 notebooks, please use import style 
  discussed in https://git.io/fecon-intro 

* Docker container (optional): instantly run fully-configured programs and 
  interactive notebooks; start by: `docker pull rsvp/fecon235` # see our 
  [Docker] image for details.  

Some basic commands like `get()` and `plot()`
will do a lot of the heavy lifting to get you started immediately. 
The commands are very easy to customize, 
producing sophisticated results quickly 
without tweaking the underlying numerical packages.

The *docs* directory should be gradually adding tutorials. 
The source code, in the meantime, is thoroughly annotated.


### Examples of code

The best way to see the code in action is to 
run the notebooks in the `nb` directory
which are described further below. 
Note that GitHub can render Jupyter notebooks directly in the browser, 
however, they will not be executable.

Here is a rendering of a notebook at GitHub for 
*Housing economy, home prices and affordibility* https://git.io/housing 
If you had executed that notebook locally, it would 
have also retrieved the latest available data and 
recomputed the results.

How is worker's wage correlated with GDP output? See https://git.io/gdpwage
How much Federal debt must each worker assume? And how fast is the
US government debt increasing? https://git.io/debtpop

To score the Federal Reserve's performance under its dual
mandate for inflation and unemployment, see https://git.io/fed
(where tangentially the Phillips curve is discredited by constructing
heat map scatter plots). 
Notebook https://git.io/infl gives an in-depth analysis of inflation,
including a combined forecast using three orthogonal methods.

Elevated default risk across bond markets is indicative of a
weak economy. But how can a policy maker calibrate credit spreads
to assess changes in interest rates? We consider mortgage and corporate
credit spreads to construct a *robust* Unified Credit Profile
(a tutorial on MAD, Median Absolute Deviation, in
rescaling non-Gaussian time-series), see https://git.io/creditprof

Please see https://git.io/fedfunds 
to forecast the Fed Funds rate using futures contracts on LIBOR.

The notebook https://git.io/cotr discerns how various asset classes 
are positioned in the market. In contrast, an overview of asset prices is 
given in https://git.io/georet using geometric mean returns.

In https://git.io/gold we make a conjecture that 
real gold prices is a stationary time-series bound by real interest rates.
In https://git.io/xbt Bitcoin is statistically analyzed as a financial asset.
We examine the crude oil markets, specifically the Brent over WTI spread,
and construct an optimal portfolio, in https://git.io/oil .

SEC 13F filings can be easily parsed, see https://git.io/13F
where we track asset managers Stanley Druckenmiller and John Paulson.

In https://git.io/equities we examine the separable components
of total return for equities, especially due to enterprise earnings
and market speculation, using S&P data assembled by Robert Shiller
which goes back to the year 1871.
In https://git.io/gdpspx we examine the close relationship between the
real economy and the equities market, while demonstrating
the Holt-Winters time-series model for predictions.

In https://git.io/gmix we analytically and visually show how a Gaussian
Mixture model handles "fat tail" risk of leptokurtotic financial assets
under small-sample conditions.
Markowitz portfolios, designed in the arithmetic mean-variance framework
for a static period, are notoriously fragile when markets change.
In contrast, our Boltzmann portfolios are adaptive over multi-periods to
*geometrically* maximize wealth using techniques from reinforcement learning.
Part 1: https://git.io/boltz1 Part 2: https://git.io/boltz2

Bootstrapping has two benefits: small-sample statistics
and simulation from controlled population.
In https://git.io/bootspx we simulate alternate histories
for leptokurtotic SPX equity returns:
to visualize sample price paths, and for
estimating probabilities of events such as investment loss.


### Development and contacts ###

* Guidelines: we welcome your [pull request] to improve our code. 
  Details are outlined in [Development].

* Lead developer is Adriano [rsvp.github.com][admin].
  Please join our chat with fellow users and developers at [Gitter].

This project is a derivative from the seminar series held at the
University of California at Berkeley, jointly sponsored by the
Department of Economics and the Haas School of Business.
We are also grateful to [BIDS], Berkeley Institute for Data Science,
and the [Mathematical Sciences Group][MathSci] for their technical support.


![fecon235-wordclouds.jpg](https://git.io/fecon235words)


### Partial contents of nb directory ###

##### [gauss-mix-kurtosis.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/gauss-mix-kurtosis.ipynb) : Gaussian Mixture and Leptokurtotic Assets

Gaussian Mixture GM(n) can create distributions with leptokurtosis ("fat
tails"). Our problem is the inverse: from observable statistics, deduce the
model parameters analytically. We demonstrate how a GM(2) model can
synthesize Gaussian risk-equivalence for leptokurtotic financial assets.
A numerical solution provides accurate probabilities which can be used to
experimentally understand how kurtosis itself is distributed under
small-sample conditions. The non-Gaussian distributions are visualized through
quantile-quantile probability plots. Shortcut: https://git.io/gmix

##### [boots-eq-spx.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/boots-eq-spx.ipynb) : Bootstrap leptokurtotic SPX equity returns

Bootstrapping has two benefits: small-sample statistics
and simulation from controlled population.
We simulate alternate histories: to visualize sample price paths,
and for estimating probabilities of events such as investment loss.
Bootstrapping facilitates study of small-sample behaviour for which
asymptotic statistical theory is unsuitable, or where closed-form
mathematical analysis is intractable, for example, *geovolatility*
which is the volatility of the geometric mean rate.
Shortcut: https://git.io/bootspx

##### [prtf-boltzmann-1.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/prtf-boltzmann-1.ipynb) : Boltzmann portfolios

We develop an alternative to the Markowitz framework
called Boltzmann portfolios which handle uncertainty from the 
standpoint of cross-entropy and optimal sequential decisions.
The improved result is a faster online algorithm which is more robust.
Markowitz portfolios are designed in the arithmetic mean-variance framework
for a static period, and are fragile to changing market conditions.
In contrast, Boltzmann portfolios are adaptive over multi-periods to
geometrically maximize wealth using techniques from reinforcement learning.
Part 1: https://git.io/boltz1 Part 2: https://git.io/boltz2

##### [qdl-spx-earn-div.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/qdl-spx-earn-div.ipynb) : Separable components of total return for equities

We specify a model for equity returns by decomposition into enterprise and
speculative returns, plus dividend yield. That model is then tested using stock market
data going back to the year 1871 (well-known database assembled by Robert Shiller).
An understanding of their respective contributions helps us to form
better informed expectations of total return for equities.
We demonstrate that the (arithmetic) percentage reasoning is prone is large errors,
whereas a logarithmic (geometric) version is exact.
Shortcut: https://git.io/equities or https://git.io/spx

##### [qdl-libor-fed-funds.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/qdl-libor-fed-funds.ipynb) : Use pandas to analyze short-term rates

We examine the spread between two interest rates: LIBOR and Fed Funds. 
The former has a much greater depth in the futures market in terms of 
volume and maturity horizon, implying richer information content. 
Modeling their relationship, we construct a synthetic forward Fed Funds 
rate, useful in gauging market sentiment regarding Fed policy. 
Estimate is given for the change in Fed Funds rate over the 
next 12 months. Shortcut: https://git.io/fedfunds

##### [qdl-xau-contango.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/qdl-xau-contango.ipynb) : Use pandas to analyze gold contango

The *London Bullion Market Association* ceased publishing daily data 
on their *Gold Forward Offered Rate* (**GOFO**), as of 30 January 2015 -- 
so we develop an observable proxy called *tango* using gold futures 
and LIBOR. This supply/demand indicator is then compared against 
change in spot prices. Observed strong correlations appear 
to be artificial. Shortcut: https://git.io/xau-contango

##### [qdl-xbt-bitcoin.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/qdl-xbt-bitcoin.ipynb) : Statistical analysis of Bitcoin as financial asset

We first examine time-series data for price, mining, and capitalization of Bitcoin, 
then optimize a robust model for the extremely volatile USD price series.
Taking the viewpoint of a Chinese user we perform a comparative valuation in 
Chinese yuan, and also cross-check with the perennial store of value: gold.
The astonishing volatility and geometric return makes Bitcoin a 
speculative financial asset which may hinder it as a payment system.
Shortcut: https://git.io/xbt

##### [qdl-COTR-positions.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/qdl-COTR-positions.ipynb) : Use pandas to read CFTC COTR

Commitment of Traders Report (COTR) is useful to extract market 
positions in precious metals, US dollar, interest rates, and 
equities markets. We develop our own scale-free measures to 
gauge market sentiment across time which can 
diverge from price directionality at interesting points.
Shortcut: https://git.io/cotr

##### [SEC-13F-parse.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/SEC-13F-parse.ipynb) : Use pandas to read 13F filings from SEC

Sort percentage allocation to long equities. 
Caveats are noted for portfolio management. 
Module yi_secform easily sums up 13F filings by one function.
For illustration, we follow asset managers with significant positions in GLD,
a gold ETF; see Stanley Druckenmiller's sudden accumulation,
and John Paulson's dramatic liquidation.
Shortcut: https://git.io/13F

##### [fred-credit-spreads.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-credit-spreads.ipynb) : Robust profiles of credit spreads

Serves as a tutorial on MAD, Median Absolute Deviation,
for robustly rescaling non-Gaussian time-series.
We consider mortgage and corporate credit spreads to construct a
robust Unified Credit Profile to calibrate credit default risk in
the context of monetary policy. Shortcut: https://git.io/creditprof

##### [fred-debt-pop.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-debt-pop.ipynb) : Growth of Federal debt, its burden on the US population  

We examine US government debt in real terms, and the current Federal debt per capita. 
Shortcut: https://git.io/debtpop

##### [fred-employ-nfp.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-employ-nfp.ipynb) : US employment data, Nonfarm Payroll

We focus on forecasting the monthly change in NFP using a variety of optics:
baseline expectation since 1939, Holt-Winters method, visual selection of local range,
regression against economic activity (SPX) -- but the
standard errors are inherently very large due to survey measurement error.

##### [fred-eur-fx.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-eur-fx.ipynb) : Euro currency qua Foreign Exchange  

We examine euro FX data from the Fed Reserve FRED database. Our synthetic
time-series, which takes us far back as 1971, give additional perspective to
observe the cross-rates against U.S. dollar and Japanese yen.  

##### [fred-eurozone.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-eurozone.ipynb) : Eurozone economics  

We examine the usual suspects: unemployment, inflation, real interest rate,
foreign exchange rate, comparative GDP. Appendix 1 concisely explains the
*euro crisis* in a video.  

##### [fred-gdp-spx.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-gdp-spx.ipynb) : US real GDP vs. SPX: Holt-Winters time series forecasting  

We examine the US gross domestic product's relationship to the US equity
market (S&P 500), in real terms. Forecasts for both are demonstrated 
using the **Holt-Winters time-series model**. We derive the most likely range 
for real GDP growth, and identify extreme equity valuations aside from 
inflationary pressures. Shortcut: https://git.io/gdpspx

##### [fred-gdp-wage.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-gdp-wage.ipynb) : U.S. GDP vs. Wage Income 

How is wage correlated with GDP output?  In answering this question,
we derive a model for GDP growth based on observations from wage growth. 
Shortcut: https://git.io/gdpwage

##### [fred-georeturns.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-georeturns.ipynb) : Comparative geometric mean returns

We examine economic and financial time series where Holt-Winters is used to
forecast one-year ahead. Daily data for bonds, equity, and gold is then
analyzed. The focus is on geometric mean returns because they optimally
express mean-variance under logarithmic utility. Shortcut: https://git.io/georet

##### [fred-housing.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-housing.ipynb) : Housing economy, home prices and affordibility  

Alan Greenspan in 2014 pointed out that there was never a recovery from recession
without improvements in housing construction. Here we examine some relevant
data, including the Case-Shiller series, and derive an insightful measure of
the housing economy, *hscore*, which takes affordibility into account.
Shortcut: https://git.io/housing

##### [fred-inflation.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-inflation.ipynb) : Inflation data from FRED using pandas

We examine inflation data: CPI and PCE, including the core versions, along
with the 10-year BEI rate (break-even inflation) from the bond market.
A unified inflation statistic *m4infl* is introduced,
which leads to the estimation of the geometric mean rate.
A robust optimized Holt-Winters model is used for forecasting.
Shortcut: https://git.io/infl

##### [fred-infl-unem-fed.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-infl-unem-fed.ipynb) : Score for the Fed's dual mandate

We examine unemployment and inflation data to construct a time-series which
gives a numerical score to the Fed's performance on its dual mandate. 
The key is to find comparable units to measure performance and a suitable
scalar measure to show deviation from the dual mandate. The visualization
includes sequential scatter plots using color heat map, which can be 
extended to studies of the Phillips curve. Shortcut: https://git.io/phillips

##### [fred-infl-velocity.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-infl-velocity.ipynb) : Inflation, money velocity, and interest rates  

We examine and attempt to replicate the results of two interesting articles by
Yi Wen and Maria Arias -- along the way, we take a detour and explore the
connection between money velocity and bond rates. This will tie together their
relationship with GDP and the money supply in a fitted equation.  

##### [fred-oil-brent-wti.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-oil-brent-wti.ipynb) : Oil: Brent vs. West Texas Intermediate (WTI)

We examine the history of crude oil prices, and their spreads.
A Boltzmann portfolio is computed for *optimal* financial positions.
Deflated prices give additional insight, along with some of the statistical
tools useful in financial economics.
Although WTI is more desirable than Brent from a petrochemical perspective,
that preference is reversed when the metrics are financial.
Shortcut: https://git.io/oil

##### [fred-usd-RTB-xau.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-usd-RTB-xau.ipynb) : Real trade-weighted indexes for USD, gold, and SPX  

We examine the value of USD against a basket of 26 foreign currencies using
real trade numbers. Trade statistics are released annually, however, the Fed
uses international inflation data to adjust the weights monthly.  

##### [fred-wage-capital.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-wage-capital.ipynb) : Real capital equivalence to wage-income 

We determine how much real capital has been necessary for risk-free interest
to match annual wage.  

##### [fred-xau-spx.ipynb](https://github.com/rsvp/fecon235/blob/master/nb/fred-xau-spx.ipynb) : Gold vs. SP500 returns, XAU vs. SPX  

Long-term comparison of two asset classes: we boxplot their return
distribution and also compute geometric mean returns. Correlation between the
two is shown to be nil. We then look at the history of projected returns using
Holt-Winters method, which also gives the latest forecasts. To conclude, we

CHANGELOG.md
## CHANGE LOG 

*Each file in this project generally has a detailed change log contained 
within itself. This file simply gives a grand overview of such details 
and the annotations in the commits and tags.*


### 2018-12-03  (tag: v6.18.1203)

Add nb/boots_ndl_d4spx_1957-2018.csv.gz and nb/boots-eq-spx.ipynb
for Bootstrapping, which demonstrates fecon236 sim and bootstrap
modules, applied to leptokurtotic SPX equity returns.

Add nb/fred-credit-spreads.ipynb which also serves as a
tutorial on MAD, Median Absolute Deviation,
in robustly rescaling non-Gaussian time-series.
We consider mortgage and corporate credit spreads to
construct a robust Unified Credit Profile to calibrate
credit default risk in the context of monetary policy.


### 2018-06-23  (tag: v6.18.0623)

**Major version change from v5 to v6**
signaling our integration with **fecon236**
which was spun-off from our source code.

**Henceforth, fecon235 becomes a repository solely of Jupyter notebooks.**
The old Python source code at fecon235 will remain for archival purposes,
while new code development shifts over to fecon236.
Please see https://git.io/econ for details.

Refresh README.md, esp. "What for" section,
include "Migration notice" at the top
regarding separation of tools from notebooks.

Revise docs/fecon235-00-README.ipynb to introduce fecon236.
Function names have been retained, but under fecon236
the call routing is expected to be more explicit than casual,
i.e. modules names are more significant.


### 2018-03-12  (tag: v5.18.0312)

Update and fix fred-oil-brent-wti.ipynb per issue #2.
New shortcut: https://git.io/oil for crude oil markets.
New introductions regarding petrochemicals
and the Boltzmann portfolio of oils.

nb/fred-debt-pop.ipynb: Update preamble and data.
Use gemrat() instead of georet().
Append Appendix 1. New shortcut: https://git.io/debtpop

fecon235.py: Add foreholt() function which
supercedes yi_fred.holtfred(), but
move holtfred() to fecon235 module for backward compatibility.
Then at lib/yi_fred.py: comment out holtfred() with notification.
We only need to maintain foreholt().
Pre-2016 notebooks with updated import and preambles should work.

Fix fred-inflation.ipynb per issue #2, and optimize forecast.
Holt-Winters model now uses robust optimized parameters.
Unified inflation has its own section.
Geometric mean rate introduced.
Drop section on correlation to gold.
Summary combines three orthogonal forecasts.

fecon235.py: Add foreinfl() to forecast Unified Inflation.
The best documentation for this function is nb/fred-inflation.ipynb
which shows how it was derived by interacting with data and plots.
This single function distills the forecasting process
derived in the notebook. Further discussed in new Appendix 2.


### 2017-07-22  (tag: v5.17.0722)

lib/ys_gauss_mix.py: Add gm2gemrat() and gm2gem().
Unify GM(2) model and gemrat() with only one pass through data.
Fallback clause for gemrate() adopts
second-order approximation used in georet().
Note: nan will occur when expected losses exceed 100% -- log error!!
Such mean estimates actually occurred during 2008Q4, Great Recession.

nb/fred-georeturns.ipynb: Replace groupgeoret() by groupgemrat().

Add docs/fecon235-08-sympy.ipynb: SymPy tutorial.
Demo sympy with LaTeX features.
Reference its use with Gaussian mixture models.

Add lib/yi_matrix.py: Linear algebra module.
Numerically understand numpy inverse methods.
Add cov2cor(): convert covariance to correlation coefficients.

Add lib/ys_mlearn.py: new softmax().
New module for machine learning tools.
Function softmax() is used in cross-entropy and MLE situations,
useful for constructing Boltzmann portfolios.
New softmax_sort(): sort on probabilities with options
for filter and renormalization.

Add lib/ys_prtf_boltzmann.py: Boltzmann portfolio,
alternative to Markowitz portfolio.
Usage demonstrated in notebook nb/prtf-boltzmann-1.ipynb

Add nb/prtf-boltzmann-2.ipynb:
Investigate temperature parametization, portfolio weight dynamics.
Add Appendix 1 regarding August 2011 global crisis
and learning from stress conditions.


### 2017-06-03  (tag: v5.17.0603)

lib/yi_0sys.py: Add timestamp() per RFC-3339 standard.

Relationship between the real economy and the equities market,
fred-gdp-spx.ipynb: Fix issue #2, optimize HW parameters,
update data and narrative.
Usage of the code for the Holt-Winters time-series model is illustrated
in the Jupyter notebook rendered at https://git.io/gdpspx

lib/yi_1tools.py: Add toar(), converter to pure array.
Add df2a() to convert single column dataframe to np array
of shape (n,) rather than (n,1) which can be annoying.
Add pastear() to merge arrays as columns.
Add diflog(), and accept numpy array as data;
this takes the difference between lagged log(data).

lib/yi_1tools.py: Add roundit() to round floats from iterable.
Echo or return a list from iterable where floats are rounded n places.
This is mainly to make some lengthy output readable.

lib/yi_1tools.py: Add kurtfun() to compute Pearson kurtosis, and
append said function to stat().

lib/yi_simulation.py: add random functions:
uniform randou(), and indicator maybe().
Add Gaussian randog(), simug(), and simug_mix().
Mathematical usage given in nb/gauss-mix-kurtosis.ipynb
which covers Gaussian mixture GM(n) models.

gauss-mix-kurtosis.ipynb: Analytic solution for GM(2)
is complete. Use sympy to compute model parameters numerically.

Fix gm2_strategy() when inadmissible solution occurs.
A seemingly free parameter like b may not lead to a solution
in the real domain, but a slight upward adjustment may help
to avoid the imaginary domain, thanks to sympy.

lib/yi_plot.py: Add plotqq() for Q-Q probability plot.
Quantile-quantile plots are used to compare data
to a theoretical distribution (default is Gaussian).

lib/ys_gauss_mix.py: Refine geometric mean approximation.
Add gemreturn_Jean(), gemrate(), and for data: georat().
These involve kurtosis, but not the GM(2) sigma decomposition.


lib/yi_0sys.py
#  Python Module for import                           Date : 2017-05-15
#  vim: set fileencoding=utf-8 ff=unix tw=78 ai syn=python : per Python PEP 0263 
''' 
_______________|  yi_0sys.py : system and date functions including specs.

Code in this module must be compatible with both Python 2 and 3.
It is a bridge and a guardian between the two Pythons.

For example, it is used in the preamble of fecon235 Jupyter notebooks.


REFERENCES:
- Compatible IDIOMS: http://python-future.org/compatible_idioms.html
                     Nice presentation.

- SIX module is exhaustive: https://pythonhosted.org/six/
        Single file source: https://bitbucket.org/gutworth/six


CHANGE LOG  For latest version, see https://github.com/rsvp/fecon235
2017-05-15  Add timestamp() per strict RFC-3339 standard.
               Also include scipy and sympy in specs().
2017-03-16  Revise: minimumPandas = 18.0
2016-04-28  Use version('jupyter_core') instead of version('jupyter').
               Update to PREAMBLE-p6.16.0428
2015-12-29  For errf in gitinfo(), our dev_null instead of os.devnull
               Add minimumPandas variable esp. for tests.
2015-12-27  Get jupyter version among specs().
               Fix run command to accept errf argument.
               For specs(), move gitinfo try clause therein.
2015-12-23  Add run command and gitinfo functions.
               Update to PREAMBLE-p6.15.1223
2015-12-19  python3 compatible: absolute_import
2015-12-03  First version.
'''

from __future__ import absolute_import, print_function
#    __future__ for Python 2 and 3 compatibility; must be first in file.
import sys
import os
import time
from subprocess import check_output, STDOUT
#                      ^for Python 2.7 and 3+


minimumPython = ( 2, 7, 0 )
#             ... else a warning is generated in specs().
minimumPandas = 18.0
#               ^after replacing the first dot, then float.


#             Open  /dev/null equivalent file for Unix and Windows:
dev_null = os.open(os.devnull, os.O_RDWR)
#                                 ^Cross-platform read and write. 
#             os.devnull is just a string:  "/dev/null" or "nul"
#             thus redirecting to os.devnull is insufficient
#             and that alone will cause a fileno error.
#  We could later close it by: os.close(dev_null). Leave open.
#  See gitinfo() for example of usage.


def getpwd():
    '''Get present working directory (Linux command is pwd).
       Works cross-platform, giving absolute path.
    '''
    return os.getcwd()


def program():
    '''Get name of present script; works cross-platform.'''
    #  Note: __file__ can get only the name of this module.
    return os.path.basename(sys.argv[0])


def warn( message, stub="WARNING:", prefix=" !. "):
    '''Write warning solely to standard error.'''
    print(prefix, stub, program(), message, sep=' ', file=sys.stderr)


def die( message, errcode=1, prefix=" !! "):
    '''Gracefully KILL script, optionally specifying error code.'''
    stub = "FATAL " + str(errcode) + ":"
    warn( message, stub, prefix )
    sys.exit( errcode )
    #         ^interpretation is system dependent;
    #          generally non-zero is considered as some error.
    #  Note: "os._exit" exits without calling cleanup handlers, 
    #  flushing stdio buffers, etc. Thus, it is not a standard way.


def date( hour=True, utc=True, localstr=' Local' ):
    '''Get date, and optionally time, as ISO string representation.
       Boolean hour variable also gives minutes and seconds.
       Setting utc to False will give local time instead of UTC,
       then localstr can be used to indicate location.
    '''
    if hour:
        form = "%Y-%m-%d, %H:%M:%S"
    else:
        form = "%Y-%m-%d"
    if utc:
        form += ' UTC'
        tup = time.gmtime()
    else:
        form += localstr
        tup = time.localtime()
    return time.strftime( form, tup ) 


def timestamp():
    '''Timestamp per strict RFC-3339 standard where timezone Z:=UTC.'''
    form = "%Y-%m-%dT%H:%M:%SZ"
    tup = time.gmtime()
    return time.strftime( form, tup ) 


def pythontup():
    '''Represent invoked Python version as an integer 3-tuple.'''
    #  Using sys.version is overly verbose.
    #  Here we get something like (2, 7, 10) which can be compared.
    return sys.version_info[:3]


def versionstr( module="IPython" ):
    '''Represent version as a string, or None if not installed.'''
    #  Unfortunately must treat Python vs its modules differently...
    if module=="Python" or module=="python":
        ver = pythontup()
        return str(ver[0]) + '.' + str(ver[1]) + '.' + str(ver[2])   
    else: 
        try:
            exec( "import " + module )
            exec( "vermod = " + module + ".__version__" )
            return vermod
        except:
            return None


def versiontup( module="IPython" ):
    '''Parse version string into some integer 3-tuple.'''
    s = versionstr(module)
    try:
        v = [ int(k) for k in s.split('.') ]
        return tuple(v)
    except:
        #  e.g. if not installed or not convertible to integers...
        if s == None:
            return ( 0,  0,  0)
        else:
            return (-9, -9, -9)


def version( module="IPython" ):
    '''Pretty print Python or module version info.'''
    print(" :: ", module, versionstr(module))


def utf( immigrant, xnl=True ):
    '''Convert to utf-8, and possibly delete new line character.
       xnl means "delete new line"
    '''
    if xnl:
        #                Decodes to utf-8, plus deletes new line.
        return immigrant.decode('utf-8').strip('\n')
    else:
        #                Decode for compliance to utf-8:
        return immigrant.decode('utf-8')


def run( command, xnl=True, errf=None ):
    '''RUN **quote and space insensitive** SYSTEM-LEVEL command.
       OTHERWISE: use check_output directly and list component 
       parts of the command, e.g.  
          check_output(["git", "describe", "--abbrev=0"])
       then generally use our utf() since check_output 
       usually does not return utf-8, so be prepared to 
       receive bytes and also new line.
    '''
    #  N.B. -  errf=None means the usual error transmittal.
    #          Cross-platform /dev/stdout is STDOUT
    #          Cross-platform /dev/null   is our dev_null above.
    #  https://docs.python.org/2/library/subprocess.html
    return utf( check_output(command.split(), stderr=errf), xnl )



def gitinfo():
    '''From git, get repo name, current branch and annotated tag.'''
    #  Suppressing error messages by os.devnull seems cross-platform,
    #  but it is just a string, so use our open file dev_null instead.
    try:
        repopath = run("git rev-parse --show-toplevel", errf=dev_null)
        #              ^returns the dir path plus working repo name.
        repo = os.path.basename(repopath)
        tag = run("git describe --abbrev=0", errf=dev_null)
        #                       ^no --tags because we want annotated tags.
        bra = run("git symbolic-ref --short HEAD", errf=dev_null)
        #         ^returns the current working branch name.
        return [repo, tag, bra]
    except CalledProcessError:
        #  Probably outside git boundaries...
        return ['git_repo_None', 'tag_None', 'branch_None']


def specs():
    '''Show ecosystem specifications, including execution timestamp.'''
    #  APIs are subject to change, so versions are critical for debugging:
    version("Python")
    if pythontup() < minimumPython:
        warn("may need newer Python version.")
    version("IPython")
    version("jupyter_core")
    version("notebook")
    #       ^worked for Jupyter notebook 4.0.6
    version("matplotlib")
    version("numpy")
    #       ^dependency for pandas
    version("scipy")
    version("sympy")
    version("pandas")
    version("pandas_datareader")
    #       ^but package is "pandas-datareader" esp. for financial quotes. 
    repo, tag, bra = gitinfo()
    print(" ::  Repository:", repo, tag, bra )
    print(" ::  Timestamp:", timestamp() )


if pythontup() < (3, 0, 0):
    '''ROSETTA STONE FUNCTIONS approximately bridging Python 2 and 3.
    e.g.       answer = get_input("Favorite animal? ")
               print(answer)
    '''
    get_input = raw_input
else:
    get_input = input
    #           ^beware of untrustworthy arguments!


def endmodule():
    '''Procedure after __main__ conditional in modules.'''
    die("is a MODULE for import, not for direct execution.", 113)


if __name__ == "__main__":
    endmodule()



'''
_______________ Appendix 1: PREAMBLE for Jupyter NOTEBOOKS
                            Input cell for settings and system details:

CHANGE LOG
2016-04-28  Remove old LaTeX warnings:
            #  Beware, for MATH display, use %%latex, NOT the following:
            #                   from IPython.display import Math
            #                   from IPython.display import Latex


#  PREAMBLE-p6.16.0428 :: Settings and system details
from __future__ import absolute_import, print_function
system.specs()
pwd = system.getpwd()   # present working directory as variable.
print(" ::  $pwd:", pwd)
#  If a module is modified, automatically reload it:
%load_ext autoreload
%autoreload 2
#       Use 0 to disable this feature.

#  Notebook DISPLAY options:
#      Represent pandas DataFrames as text; not HTML representation:
import pandas as pd
pd.set_option( 'display.notebook_repr_html', False )
from IPython.display import HTML # useful for snippets
#  e.g. HTML('<iframe src=http://en.mobile.wikipedia.org/?useformat=mobile width=700 height=350></iframe>')
from IPython.display import Image 
#  e.g. Image(filename='holt-winters-equations.png', embed=True) # url= also works
from IPython.display import YouTubeVideo
#  e.g. YouTubeVideo('1j_HxD4iLn8', start='43', width=600, height=400)
from IPython.core import page
get_ipython().set_hook('show_in_pager', page.as_hook(page.display_page), 0)
#  Or equivalently in config file: "InteractiveShell.display_page = True", 
#  which will display results in secondary notebook pager frame in a cell.

#  Generate PLOTS inside notebook, "inline" generates static png:
%matplotlib inline   
#          "notebook" argument allows interactive zoom and resize.


'''

